### PR TITLE
send back Widget class name information in the widget rebuild/repaint count events

### PR DESCRIPTION
**Pull request by devoncarew**
This PR:
- adds tests to validate that we're collecting Widget class names in the widget kernel transformer
- send back Widget class name information in the widget rebuild/repaint count events

*List which issues are fixed by this PR. You must list at least one issue.*

- https://github.com/flutter/devtools/issues/3143

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by devoncarew** [comment]
I encoded the names in a new `'newLocationsNames'` field in the returned event, but there might be other ways to encode the information.

The location is, line, and column are encoded in a `List<int>` in the `'newLocations'` field, so would be challenging to include the name info (a String) there.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 1]
```diff
@@ -2137,6 +2137,25 @@ class _ElementLocationStatsTracker {
       }
       json['newLocations'] = locationsJson;
     }
+
+    // Add in a data structure for the location names.
+    if (newLocations.isNotEmpty) {
+      final Map<String, Map<int, String>> namesJson = <String, Map<int, String>>{};
+      for (final _LocationCount entry in newLocations) {
+        final _Location location = entry.location;
+        if (location.file != null) {
```
Comment (jacob314) [inline packages/flutter/lib/src/widgets/widget_inspector.dart:2146]: as a follow up, location.file can now be non-nullable so you can delete this `!= null` check and others in the file.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 2]
```diff
@@ -2137,6 +2137,25 @@ class _ElementLocationStatsTracker {
...
+          final Map<int, String> jsonForFile = namesJson.putIfAbsent(
+            location.file!,
+                () => <int, String>{},
+          );
+          if (location.name != null) {
```
Comment (jacob314) [inline packages/flutter/lib/src/widgets/widget_inspector.dart:2151]: nit:
add 
`final name = location.name;`
to eliminate the `!` on the following line.

File: packages/flutter/test/widgets/widget_inspector_test.dart [hunk 3]
```diff
@@ -943,6 +945,8 @@ class _TestWidgetInspectorService extends TestWidgetInspectorService {
       final String fileA = creationLocationA['file']! as String;
       final int lineA = creationLocationA['line']! as int;
       final int columnA = creationLocationA['column']! as int;
+      final String? nameA = creationLocationA['name'] as String?;
```
Comment (jacob314) [inline packages/flutter/test/widgets/widget_inspector_test.dart:948]: nit: keep line 948 consistent with line 945. Either the new way or the old way of reading from the json is reasonable.

**Review by jacob314 — APPROVED** [review]
lgtm

File: packages/flutter/lib/src/widgets/widget_inspector.dart (hunk 1 above)
Comment (devoncarew) [reply packages/flutter/lib/src/widgets/widget_inspector.dart:2146]: done

File: packages/flutter/lib/src/widgets/widget_inspector.dart (hunk 2 above)
Comment (devoncarew) [reply packages/flutter/lib/src/widgets/widget_inspector.dart:2151]: done

File: packages/flutter/test/widgets/widget_inspector_test.dart (hunk 3 above)
Comment (devoncarew) [reply packages/flutter/test/widgets/widget_inspector_test.dart:948]: done

**Pull request merged**